### PR TITLE
Fix retry logic for git clone

### PR DESCRIPTION
**Pull request by baierjan**
References:
- https://progress.opensuse.org/issues/123873
- https://progress.opensuse.org/issues/123556
- https://progress.opensuse.org/issues/123888

### CODE REVIEW

File: OpenQA/Isotovideo/Utils.pm [hunk 1]
```diff
@@ -101,10 +104,18 @@ sub checkout_git_repo_and_branch ($dir_variable, %args) {
     my $clone_url = $url->fragment(undef)->to_string;
     my $local_path = $url->path->parts->[-1] =~ s/\.git$//r;
     my $tries = $args{retry_count};
-    while ($tries--) {
-        last unless clone_git $local_path, $clone_url, $args{clone_depth}, $branch, $dir, $dir_variable;
-        bmwqemu::diag "Clone failed, retries left: $tries of $args{retry_count}";
-        sleep($args{retry_interval} // 5);
+  retry: {
```
Comment (Martchus) [inline OpenQA/Isotovideo/Utils.pm:107]: Where is this label used?

File: OpenQA/Isotovideo/Utils.pm [hunk 2]
```diff
@@ -101,10 +104,18 @@ sub checkout_git_repo_and_branch ($dir_variable, %args) {
...
+        do {
+            my $status;
+            eval { $status = clone_git $local_path, $clone_url, $args{clone_depth}, $branch, $dir, $dir_variable };
+            $error = $@;
+            last if $status;
+            bmwqemu::diag "Clone failed, retries left: $tries of $args{retry_count}";
+            rmdir $local_path;
+            sleep($args{retry_interval} // 5);
+        } while ($tries-- > 0);
+        die $error;
```
Comment (Martchus) [inline OpenQA/Isotovideo/Utils.pm:118]: This `die` is outside of the loop. Wouldn't we always run into this case?

**Review by Martchus — CHANGES_REQUESTED** [review]
Even if this code makes sense it is perhaps very hard to follow.

File: OpenQA/Isotovideo/Utils.pm (hunk 2 above)
Comment (baierjan) [reply OpenQA/Isotovideo/Utils.pm:118]: No, we hit it only if the `last` is not used in the inner loop. The `last` will end the named block, not the do-while.

Comment (Martchus) [reply OpenQA/Isotovideo/Utils.pm:118]: Ok, this makes sense. I'm still wondering whether we can simplify this. Having a label that is not really referenced and having two nested loops seems a bit too complicated for this use-case.

Comment (baierjan) [reply OpenQA/Isotovideo/Utils.pm:117]: I really tried, this construction was suggested on https://www.perltutorial.org/perl-do-while/

Comment (baierjan) [reply OpenQA/Isotovideo/Utils.pm:117]: I can probably use `last retry if $status;` if you want the label to be mentioned somewhere.

Comment (okurz) [reply OpenQA/Isotovideo/Utils.pm:117]: Would it work to put this code into a sub-function and return from there instead of a loop with label?

**Review by okurz — COMMENTED** [review]
Please describe in both commits a bit more explanation. I wonder why you turn of signing. Sounds useful to do but I'd still like to learn that brought you to this. And also for the retry commit

File: OpenQA/Isotovideo/Utils.pm (hunk 2 above)
Comment (baierjan) [reply OpenQA/Isotovideo/Utils.pm]: I managed to rearrange the code slightly so the second loop is no longer needed. It behave slightly differently (setting `$bmwqemu::vars{$dir_variable}` before it dies)  , hopefully it does not matter.

**Comment by okurz** [comment]
+1, @kalikiana you might be interested as well

**Review by Martchus — APPROVED** [review]

**Review by kalikiana — APPROVED** [review]
Nice. And you improved the unit test coverage as well, I think that's a really good idea here!
